numpy/core/src/multiarray/methods.c

```diff
@@ -2694,6 +2694,185 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     return c;
 }
 
+
+static void
+array_dlpack_deleter(DLManagedTensor *self)
+{
+    PyArrayObject *array = (PyArrayObject *)self->manager_ctx;
+    // This will also free the strides as it's one allocation.
+    PyMem_Free(self->dl_tensor.shape);
+    PyMem_Free(self);
+    Py_XDECREF(array);
+}
+
+/* This is exactly as mandated by dlpack */
+static void array_dlpack_capsule_deleter(PyObject *self)
+{
+    if (PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
+        return;
+    }
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
+    if (managed == NULL) {
+        return;
+    }
+    managed->deleter(managed);
+}
+
+static DLDevice
+array_get_dl_device(PyArrayObject *self) {
+    DLDevice ret;
+    ret.device_type = kDLCPU;
+    ret.device_id = 0;
+    PyObject *base = PyArray_BASE(self);
+    if (PyCapsule_IsValid(base, NPY_DLPACK_INTERNAL_CAPSULE_NAME)) {
+        DLManagedTensor *managed = PyCapsule_GetPointer(
+                base, NPY_DLPACK_INTERNAL_CAPSULE_NAME);
+        if (managed == NULL) {
+            return ret;
```

Comment:
This leaves an exception set, which is wrong. This should either return NULL or clear the exception. How can this line even be reached?

Reply:
[`PyCapsule_GetPointer`](https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_GetPointer) can set an exception.

Reply:
And I call `PyErr_Occurred` after if needed.

Reply:
> PyCapsule_GetPointer can set an exception.

yes, but the only situation in which it does is the one the outer if protects against - so I'd suggest removing the `if` so that this path is actually taken.

> And I call PyErr_Occurred after if needed.

After what? This is a return statement, and C API functions should do exactly one of return null or set an exception

Reply:
> yes, but the only situation in which it does is the one the outer if protects against - so I'd suggest removing the `if` so that this path is actually taken.

There's no guarantee of this in the documentation that I can find that that's the only case it'll fail.

> You should explain this in the docstring.

I'll explain in the docstring.

Reply:
You're right that there is no guarantee of that sort, but you don't need such a guarantee to remove the _outer_ if statement

Reply:
Ah, I see the confusion here -- This is roughly equivalent to the following Python-syntax psuedocode:

```python
if is_numpy_capsule(self.base):  # This is what the outer if is for.
    return self.base.ptr.device
# NumPy arrays by default are on the CPU.
return (kDLCPU, 0)
```